Ask 84

okbel opened the pull request:
This PR solves issue 84
- Checkout branch ask-84  `git checkout ask-84`
- Create a Form with a Date Field
- Click Preview 
- Complete the form and test the functionality of the Date field
- Notice the Errors ( check https://github.com/coralproject/ask/issues/84 )

https://github.com/coralproject/ask/issues/84

okbel committed:
Date validation working

okbel committed:
isCompleted issue

riley commented on src/components/fields/DateField.js:
```diff
@@ -39,110 +44,110 @@ class DateField extends AskField {
-    isCompleted = !!this.state.value.length;
-    this.setState({ isValid: isValid, completed: isCompleted });
-    return !!this.props.wrapper.required ? isValid && isCompleted : isValid;
+    return !!this.props.wrapper.required ? isValid && completed : isValid;
```
you don't need the double bang here, and since there's a lot going in this ternary, please consider adding some comments explaining what's going on.

okbel replied:
Sure. It was confusing for me too. I'll add more comments to make it clearer

riley commented on src/components/fields/DateField.js:
```diff
@@ -39,110 +44,110 @@ class DateField extends AskField {
   buildValue() {
-    return this.state.month + "-" + this.state.day + "-" + this.state.year;
+    const { month, day, year } = this.state;
+    return `${month}-${day}-${year}`;
```
Should we internationalize this sort of thing?

okbel committed:
Merge branch 'master' of https://github.com/coralproject/elkhorn into ask-84

okbel committed:
Comment

okbel commented:
@riley Merge issues are solved. Also, added a comment in that ternary. For this issue I tried not to modify the logic behind it to avoid regression errors. I did some refactor and also fixed some errors while saving the completed status.

okbel committed:
Validation when its not required

okbel commented:
@riley The description of the PR has been updated. Issue 103 goes to another PR  (https://github.com/coralproject/cay/pull/674)

impronunciable commented on src/components/fields/DateField.js:
```diff
@@ -39,110 +47,111 @@ class DateField extends AskField {
-    var isValid = true;
-    var { year, month, day } = this.state;
+    this.validateAndSave()
+  }
```
missing empty line between functions

impronunciable commented on src/components/fields/DateField.js:
```diff
@@ -39,110 +47,111 @@ class DateField extends AskField {
+  }
+  isCompleted() {
+    return !!(this._year.value && this._month.value && this._day.value)
+  }
```
missing empty line between functions

impronunciable commented on src/components/fields/DateField.js:
```diff
@@ -39,110 +47,111 @@ class DateField extends AskField {
+    const { year, month, day } = this.state;
 
-    if (year < 1916 || year > 2056) {
+    if ( year && (year < 1916 || year > 2056)) {
```
extra space after `(`

impronunciable commented on src/components/fields/DateField.js:
```diff
@@ -39,110 +47,111 @@ class DateField extends AskField {
     }
 
-    if (month < 1 || month > 12) {
+    if ( month && (month < 1 || month > 12)) {
```
extra space after `(`

impronunciable commented on src/components/fields/DateField.js:
```diff
@@ -39,110 +47,111 @@ class DateField extends AskField {
-    var daysInMonth = this.daysInMonth(month, year);
-
-    if (day < 1 || day > daysInMonth) {
+    if ( day && (day < 1 || day > this.daysInMonth(month, year))) {
```
extra space after `(`

okbel committed:
Added lines between

okbel committed:
removing extra space

okbel commented:
Updated!

riley commented:
@okbel if I enter a date in only the first box, the form still lets me submit without errors. Is this expected?

https://s3.amazonaws.com/nyt-stage/57d99cf90cefe52294d0ff63.html

kgardnr commented:
@riley yeah this is expected for now. wasn't sure the best way to handle validation on date when it's not required, so only validates if they enter a full date? we can look at handling this better in the future.

riley commented:
lgtm 💯
